### ISSUE TITLE
Issue building OpenCV 3.4.3 for Android

### ISSUE THREAD

**Issue by david-serrano** [opencv/opencv#15733]
Hi,

I'm following a guide to build OpenCV 3.3.x for Android on Windows (http://amin-ahmadi.com/2017/08/13/how-to-build-opencv-3-3-x-for-android/) and I've followed all the steps in it, but the configuration fails with the following issue:

    The C++ compiler
    
        "D:/androidNdks/android-ndk-r15c/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/arm-linux-androideabi-g++.exe"
    
      is not able to compile a simple test program.
    
      It fails with the following output:
    
        Change Dir: D:/dev/opencv-3.4.3/android_build/CMakeFiles/CMakeTmp
        
        Run Build Command(s):D:/Programs/CMake/bin/cmake.exe cmTC_fc979/fast && CMake Error: The source directory "D:/dev/opencv-3.4.3/android_build/CMakeFiles/CMakeTmp/cmTC_fc979/fast" does not exist.
        Specify --help for usage, or press the help button on the CMake GUI.

I've tried with the following variations:
 - ANDROID_NDK: D/AndroidNdks/android-ndk-r15c
 - ANDROID_NDK: D/AndroidNdks/android-ndk-r16b
 - Using the cmake exe from Cmake: D:\Programs\CMake\bin
 - Using the make exe from Mingw: D:\Programs\Mingw\msys\1.0\bin
 - Updating the toolchain used from the CMake GUI
 - Swapping between Mingw Makefiles and MSYS Makefiles

I've also run CMake as an admin to make sure there are no write permission issues.

From looking online i think it's also worth noting that ma CMakeLists file is unchanged from the one downloaded from the source files ([here](https://github.com/opencv/opencv/archive/3.4.3.zip))

Any help or pointers as to what i may be doing wrong would be greatly appreciated!
I've also raised the same question on the OpenCV answers platform ([here](https://answers.opencv.org/question/219985/issue-building-opencv-343-for-android/))

Thanks, 
David

**Comment by alalek** [comment]
> OpenCV 3.4.3

Outdated version. We can't fix old releases.
You should try latest 3.4 release and latest 3.4 code before reporting.

**Comment by david-serrano** [comment]
@alalek I know it's an outdated version, but my current project inherited this and we have yet to find the time to upgrade... - seeing as the android sdk package was published for this release, im assuming it's buildable from the sources, and the issue is related to my machine/config rather than a fix required in the source!

**Comment by david-serrano** [comment]
> > OpenCV 3.4.3
> 
> Outdated version. We can't fix old releases.
> You should try latest 3.4 release and latest 3.4 code before reporting.

@alalek as per your suggestion i tried following the same tutorial steps with the latest Opencv 3.4.8 release source code. The issue i got is below:

```
Compiling the CXX compiler identification source file "CMakeCXXCompilerId.cpp" failed.
Compiler: D:/androidNdks/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/arm-linux-androideabi-g++.exe 
Build flags: -fexceptions;-frtti;-fpic;-Wno-psabi;--sysroot=D:/androidNdks/android-ndk-r16b/platforms/android-9/arch-arm;-funwind-tables;-finline-limit=64;-fsigned-char;-no-canonical-prefixes;-march=armv7-a;-mfloat-abi=softfp;-mfpu=vfpv3-d16;-fdata-sections;-ffunction-sections;-Wa,--noexecstack;;-DANDROID;-D__ANDROID_API__=9
Id flags:  

The output was:
1
D:/androidNdks/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: error: cannot open crtbegin_dynamic.o: No such file or directory
D:/androidNdks/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: error: cannot open crtend_android.o: No such file or directory
D:/androidNdks/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: error: cannot find -lstdc++
D:/androidNdks/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: error: cannot find -lm
D:/androidNdks/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: error: cannot find -lc
D:/androidNdks/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: error: cannot find -ldl
/usr/local/google/buildbot/src/android/gcc/toolchain/build/../gcc/gcc-4.9/libgcc/unwind-arm-common.inc:289: error: undefined reference to 'abort'
/usr/local/google/buildbot/src/android/gcc/toolchain/build/../gcc/gcc-4.9/libgcc/unwind-arm-common.inc:346: error: undefined reference to 'memcpy'
/usr/local/google/buildbot/src/android/gcc/toolchain/build/../gcc/gcc-4.9/libgcc/unwind-arm-common.inc:376: error: undefined reference to 'memcpy'
/usr/local/google/buildbot/src/android/gcc/toolchain/build/../gcc/gcc-4.9/libgcc/unwind-arm-common.inc:505: error: undefined reference to 'abort'
/usr/local/google/buildbot/src/android/gcc/toolchain/build/../gcc/gcc-4.9/libgcc/config/arm/pr-support.c:378: error: undefined reference to 'abort'
/usr/local/google/buildbot/src/android/gcc/toolchain/build/../gcc/gcc-4.9/libgcc/config/arm/pr-support.c:384: error: undefined reference to 'abort'
collect2.exe: error: ld returned 1 exit status
```

I then added the ANDROID_NATIVE_API_LEVEL with a value of 27 and got the following issue:
```
Determining if the CXX compiler works failed with the following output:
Change Dir: D:/dev/opencv-3.4.8/android_build/CMakeFiles/CMakeTmp

Run Build Command(s):D:/Programs/CMake/bin/cmake.exe cmTC_05c1b/fast && CMake Error: The source directory "D:/dev/opencv-3.4.8/android_build/CMakeFiles/CMakeTmp/cmTC_05c1b/fast" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
```

**Comment by david-serrano** [comment]
I fixed the build issue! (both for 3.4.3 and 3.4.8) Stupidly, CMake was not picking up the MingW Environment variable, and was also silently failing for having the MSYS Env var defined too, so after a bit of playing around with them i got it to read it and the sources built fine!
The next issue im facing is `mingw32-make` is failing to find any makefile, but that's a different story so im closing this :)
